typing/solver.ml

```diff
@@ -593,17 +788,23 @@ module Solver_mono (C : Lattices_mono) = struct
       guessed lower bound is strictly higher; and all lattices are finite.
       *)
   let zap_to_floor_morphvar_aux (type a r) (obj : a C.obj)
-      (mv : (a, allowed * r) morphvar) =
-    let rec loop lower =
+      (mv : (a, allowed * r) morphvar) : change list * a =
+    let rec loop (lower : a) : change list * a =
       let log = ref empty_changes in
-      let r = submode_mvc ~log:(Some log) obj mv lower in
+      (* Since this call to [submode_mvc] should always succeed,
+         we provide a throwaway, arbitrary hint *)
+      let r = submode_mvc ~log:(Some log) obj mv lower (Const H.const_none) in
       match r with
       | Ok () -> !log, lower
-      | Error a ->
+      | Error (a, _a_hint) ->
+        (* Note, the hint here, [_a_hint], isn't useful as it is just for explaining why
+           [a] is too high, but this is not useful as all [a] is then used for is as an
+           upper bound (not a lower bound) in the next iteration. *)
         undo_changes !log;
         loop (C.join obj a lower)
     in
-    loop (mlower obj mv)
+    let lower = mlower obj mv in
+    loop lower
```

Comment:
Since we have gone back and forth, before merging we should remove unnecessary changes like this one. 

(Just a reminder - no need to do this now :smiley: )